python/mpt/src/mpt/trie_diff.py

```diff
@@ -0,0 +1,388 @@
+
+            case (None, LeafNode()):
+                # new leaf
+                preimage = nibble_path_to_bytes(path + r_node.rest_of_key)
+                process_leaf_diff(preimage, None, r_node)
+
+            case (None, ExtensionNode()):
+                # Look for diffs in the right sub-tree
+                self._compute_diff(
+                    None, r_node.subnode, path + r_node.key_segment, process_leaf_diff
+                )
+
+            case (None, BranchNode()):
+                # Look for diffs in all branches of the right sub-tree
+                for i in range(0, 16):
+                    self._compute_diff(
+                        None,
+                        r_node.subnodes[i],
+                        path + i.to_bytes(1, "big"),
+                        process_leaf_diff,
+                    )
+
+            case (LeafNode(), None):
+                # deleted leaf (should not happen post-cancun)
+                preimage = nibble_path_to_bytes(path + l_node.rest_of_key)
+                process_leaf_diff(preimage, l_node, None)
+
+            case (LeafNode(), LeafNode()):
+                if l_node.value != r_node.value:
+                    preimage = nibble_path_to_bytes(path + l_node.rest_of_key)
+                    process_leaf_diff(preimage, l_node, r_node)
+
+            case (LeafNode(), BranchNode()):
+                # Look for diffs in all branches of the right sub-tree
+                for i in range(0, 16):
+                    if (
+                        i != l_node.rest_of_key[0]
+                        or not self._nodes[r_node.subnodes[i]].value == l_node.value
+                    ):
+                        self._compute_diff(
+                            None,
+                            r_node.subnodes[i],
+                            path + bytes([i]),
+                            process_leaf_diff,
+                        )
+
+            case (LeafNode(), ExtensionNode()):
+                # Explore the extension node's subtree for any new leaves, comparing it to the old
+                # leaf with the same key
+                l_node = LeafNode(
+                    l_node.rest_of_key[len(r_node.key_segment) :], l_node.value
+                )
+                self._compute_diff(
+                    l_node, r_node.subnode, path + r_node.key_segment, process_leaf_diff
+                )
+
+            case (ExtensionNode(), None):
+                # Look for diffs in the left sub-tree
+                self._compute_diff(
+                    l_node.subnode, None, path + l_node.key_segment, process_leaf_diff
+                )
+
+            case (ExtensionNode(), LeafNode()):
+                # The extension node was deleted and replaced by a leaf - meaning that down the line of the extension node, in a branch, we deleted some nodes.
+                # Explore the extension node's subtree for any deleted nodes, comparing it to the new leaf
+                r_node = LeafNode(
```

Comment:
here we assume that len(r_node.rest_of_key) > len(l_node.key_segment), can we add an invariant?